File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/react-router-no-empty-leaf-route.ts
import {
  REACT_ROUTER_RENDER_PROPERTY_NAMES,
  REACT_ROUTER_RESOURCE_HANDLER_PROPERTY_NAMES,
} from "../../constants/react-router.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticRouteProperty } from "../../utils/get-static-route-property.js";
import { hasActiveRouteProperty } from "../../utils/has-active-route-property.js";
import { isStaticReactRouterRouteObject } from "../../utils/is-static-react-router-route-object.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { wrapReactRouterRule } from "../../utils/wrap-react-router-rule.js";

export const reactRouterNoEmptyLeafRoute = wrapReactRouterRule(
  defineRule({
    id: "react-router-no-empty-leaf-route",
    title: "Leaf route renders nothing",
    tags: ["test-noise"],
    requires: ["react-router:6.4"],
    severity: "warn",
    recommendation:
      "Add element, Component, or lazy to a UI leaf route, or remove the empty route.",
    create: (context: RuleContext) => ({
      ObjectExpression(node: EsTreeNodeOfType<"ObjectExpression">) {
        if (!isStaticReactRouterRouteObject(context, node)) return;
        if (hasActiveRouteProperty(context, node, "children")) return;
        if (
          getStaticRouteProperty(node, "path") === null &&
          getStaticRouteProperty(node, "index") === null
        ) {
          return;
        }
        if (
          REACT_ROUTER_RENDER_PROPERTY_NAMES.some((name) =>
            hasActiveRouteProperty(context, node, name),
          )
        ) {
          return;
        }
        if (
          REACT_ROUTER_RESOURCE_HANDLER_PROPERTY_NAMES.some((name) =>
            hasActiveRouteProperty(context, node, name),
          )
        ) {
          return;
        }
        context.report({
          node,
          message:
            "This leaf route has no UI and no resource handler, so it renders a null outlet.",
        });
      },
    }),
  }),
);

File: native/oxlint/rules/react-router-no-empty-leaf-route.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    context::{ContextHost, LintContext},
    rule::Rule,
    AstNode,
};

const ROUTE_RENDER_PROPERTY_NAMES: [&str; 3] = ["Component", "element", "lazy"];
const ROUTE_RESOURCE_HANDLER_PROPERTY_NAMES: [&str; 4] =
    ["action", "clientAction", "clientLoader", "loader"];
const MESSAGE: &str =
    "This leaf route has no UI and no resource handler, so it renders a null outlet.";

#[derive(Debug, Default, Clone)]
pub struct ReactRouterNoEmptyLeafRoute;

declare_oxc_lint!(
    /// Warns when a React Router leaf route renders no UI or resource response.
    ReactRouterNoEmptyLeafRoute,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Warns when a React Router leaf route is empty.",
);

impl Rule for ReactRouterNoEmptyLeafRoute {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx) && is_react_router_file_active(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ObjectExpression(route_object) = node.kind() else {
            return;
        };
        if !is_static_react_router_route_object(route_object, ctx)
            || has_active_route_property(route_object, "children", ctx)
            || (get_static_route_property(route_object, "path").is_none()
                && get_static_route_property(route_object, "index").is_none())
            || ROUTE_RENDER_PROPERTY_NAMES
                .iter()
                .any(|property_name| has_active_route_property(route_object, property_name, ctx))
            || ROUTE_RESOURCE_HANDLER_PROPERTY_NAMES
                .iter()
                .any(|property_name| has_active_route_property(route_object, property_name, ctx))
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(route_object.span));
    }
}
